modules/core/log/src/AssetLogs.php

```diff
@@ -0,0 +1,89 @@
+<?php
+
+namespace Drupal\farm_log;
+
+use Drupal\asset\Entity\AssetInterface;
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+
+/**
+ * Service for loading logs that reference assets.
+ *
+ * @internal
+ */
+class AssetLogs implements AssetLogsInterface {
+
+  /**
+   * Entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected EntityTypeManagerInterface $entityTypeManager;
+
+  /**
+   * Log query factory.
+   *
+   * @var \Drupal\farm_log\LogQueryFactoryInterface
+   */
+  protected LogQueryFactoryInterface $logQueryFactory;
+
+  /**
+   * Class constructor.
+   *
+   * @param \Drupal\Core\Entity\EntityTypeManagerInterface $entity_type_manager
+   *   Entity type manager.
+   * @param \Drupal\farm_log\LogQueryFactoryInterface $log_query_factory
+   *   Log query factory.
+   */
+  public function __construct(EntityTypeManagerInterface $entity_type_manager, LogQueryFactoryInterface $log_query_factory) {
+    $this->entityTypeManager = $entity_type_manager;
+    $this->logQueryFactory = $log_query_factory;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getLogs(AssetInterface $asset, string $log_type = NULL, bool $access_check = TRUE): array {
+    $log_ids = $this->query($asset, $log_type, $access_check)->execute();
+    return $this->entityTypeManager->getStorage('log')->loadMultiple($log_ids);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getFirstLog(AssetInterface $asset, string $log_type = NULL, bool $access_check = TRUE) {
+    $log_ids = $this->query($asset, $log_type, $access_check, 1)->execute();
+    return $this->entityTypeManager->getStorage('log')->load(reset($log_ids));
+  }
+
+  /**
+   * Build a log query.
+   *
+   * @param \Drupal\asset\Entity\AssetInterface $asset
+   *   The asset entity.
+   * @param string|null $log_type
+   *   Optionally filter by log type.
+   * @param bool $access_check
+   *   Whether to check log entity access.
+   * @param int|null $limit
+   *   The number of logs to return.
+   *
+   * @return \Drupal\Core\Entity\Query\QueryInterface
+   *   A query object.
+   */
+  protected function query(AssetInterface $asset, string $log_type = NULL, bool $access_check = TRUE, int $limit = NULL) {
```

Comment:
Just wondering if we should make this a `public` function so you could further customize the query. It wouldn't add a whole lot... you could just use the log query factory and replicate this anyways. :shrug:

Reply:
The method can still be overridden by a downstream class (`protected` can, `private` can't) if that's what you mean? I don't think it makes sense to expose this method publicly in the `AssetLogsInterface`, since it is basically just a helper for reducing code duplication within the other two methods.

Reply:
Agree, no need to make it public then!